lib/languageFeatures/findReferencesHandler.ts

```diff
@@ -1,78 +1,119 @@
+      }
+    }
+    if (implementsIHasLexerToken(obj) && obj.lexerToken === token) {
+      definitions.add(obj);
+    }
+    if (implementsIHasEndingLexerToken(obj) && obj.endingLexerToken === token) {
+      definitions.add(obj);
+    }
+
+    if (obj instanceof OArchitecture && obj.entityName === token) {
+      if (obj.correspondingEntity) {
+        definitions.add(obj.correspondingEntity);
+      }
+    }
   }
-  if (candidate instanceof OLexerToken && candidate.parent instanceof ObjectBase) {
-    candidate = candidate.parent;
+
+  for (const definition of definitions) {
+
+    if (definition instanceof OSubprogram) {
+      definition.parent.subprograms
+        .filter(subprogram => subprogram.lexerToken.getLText() == definition.lexerToken?.getLText()).forEach(def => definitions.add(def));
+      if (definition.parent instanceof OPackage) {
+        definition.parent.correspondingPackageBodies.flatMap(packageBodies => packageBodies.subprograms
+          .filter(subprogram => subprogram.lexerToken.getLText() == definition.lexerToken?.getLText())).forEach(def => definitions.add(def));
+      }
+      if (definition.parent instanceof OPackageBody) {
+        ((definition.parent as OPackageBody).correspondingPackage?.subprograms
+          .filter(subprogram => subprogram.lexerToken.getLText() == definition.lexerToken?.getLText()) ?? []).forEach(def => definitions.add(def));
+      }
+
+
+    }
   }
-  // debugger;
-  if (candidate instanceof OReference && candidate.definitions) {
-    return candidate.definitions.concat(candidate.definitions.flatMap(c => {
-      return implementsIHasReference(c) ? c.referenceLinks : [];
-    }));
+  const definitionsNew = [...definitions].map(definition => {
+    if (definition instanceof OPackageBody && definition.correspondingPackage) {
+      return definition.correspondingPackage;
+    }
+    return definition;
+
+  });
+  const tokens: OLexerToken[] = [];
+  for (const definition of definitionsNew) {
+    if (implementsIHasReference(definition)) {
+      if (definition.lexerToken) {
+        tokens.push(definition.lexerToken);
+      }
+      tokens.push(...definition.referenceLinks.map(ref => ref.referenceToken).filter(token => token.getLText() === definition.lexerToken?.getLText()));
+      if (definition instanceof OEntity) {
+        tokens.push(...definition.correspondingArchitectures.map(arch => arch.entityName));
+        tokens.push(...definition.referenceLinks.flatMap(link => link instanceof OInstantiation ? link.componentName : []));
+      }
+      if (definition instanceof OPort) {
+        tokens.push(...definition.parent.referenceLinks
+          .flatMap(link => link instanceof OInstantiation ? link.portAssociationList?.children.flatMap(child => {
+            return child.formalPart.filter(formal => formal.referenceToken.getLText() === definition.lexerToken.getLText()).map(formal => formal.referenceToken);
+          }) ?? [] : []));
```

Comment:
This is not readable. Remove this ternary operator and/or make a proper function out of this